CDK Conceptual Docs (#3048)
airbytehq/airbyte: airbyte-integrations/bases/base-python/CDK-README.md

First is the `AirbyteStateMessage` and the `HttpStream`'s `cursor_field`. As mentioned, the `AirbyteStateMessage`
persists state between syncs, and allows a new sync to pick up from where the previous sync last finished.
The `cursor_field` refers to the actual element in the HTTP request used to determine order. The `cursor_field`'s only
practical implication is it tells the CDK the stream supports incremental syncs. Otherwise, it serves as a signpost
for what field is used as a cursor. This field is also commonly used as a direct index into the api response to
create the `AirbyteStateMessage`.

Next is the `get_updated_state` function. This function helps the CDK figure out the latest state for every record processed
(as returned by the `parse_response`function mentioned above). This allows sync to resume from where the previous sync last stopped,
regardless of success or failure. This function typically compares the state object's and the latest record's cursor field, picking the latest one.

This is paired with the `stream_slices` function. Conceptually, a Stream Slice is the smallest unit of data to be

> synced. In the HTTP case, each Slice is equivalent to a HTTP request; indeed the CDK will make one request

Reviewer: ```suggestion
synced. 

In the HTTP case, each Slice is equivalent to a HTTP request; the CDK will make one request
```

per element returned by the `stream_slices` function. A Slice object is not typed, and the developer
is free to include any information necessary to make the request. This function is called when the
`HTTPStream` is first created. Typically, the `stream_slices` function, via inspecting the state object,
generates a Slice for every request to be made.

As an example, suppose an API is able to dispense data hourly. If the last sync was exactly 24 hours ago,
we can either make an API call retrieving all data at once, or make 24 calls each retrieving an hour's
worth of data. In the latter case, the `stream_slices` function, sees that the previous state contains

> yesterday timestamp, and returns a list of 24 Slices, each with a different hourly timestamp to be

Reviewer: ```suggestion
yesterday's timestamp, and returns a list of 24 Slices, each with a different hourly timestamp to be
```

> used when creating request.

Reviewer: ```suggestion
used when creating request. If the stream fails halfway through (at the 12th slice), then the next time it starts reading, it will read from the beginning of the 12th slice. 
```

> Lastly, the `request_params` and `path` functions allow the per-request information contained in a Slice

Reviewer: ```suggestion
The current slice being read is passed into every other method in `HttpStream` e.g: `request_params`, `request_headers`, `path`, etc..
```

to be injected into a request. 

In summary, the incremental stream requires:
* the `cursor_field` property
* the `get_updated_state` function

> * the `stream_slices` function

Reviewer: ```suggestion
* Optionally, the `stream_slices` function
```

> * updating the `request_params` and `path` function to incorporate slices

Reviewer: ```suggestion
* updating the `request_params`, `path`, and other functions to incorporate slices
```

#### Secondary Features

The CDK offers other features that make writing HTTP APIs a breeze.

##### Authentication

The CDK supports token and OAuth2.0 authentication via the `TokenAuthenticator` and `Oauth2Authenticator` classes
respectively. Both authentication strategies are identical in that they place the api token in the `Authorization`
header. The `OAuth2Authenticator` goes an additional step further and has mechanisms to, given a refresh token,
refresh the current access token. Note that the `OAuth2Authenticator` currently only supports refresh tokens
and not the full OAuth2.0 loop.

Using either authenticator is as simple as passing the created authenticator into the relevant `HTTPStream`
constructor. Here is an [example](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-stripe/source_stripe/source.py#L242) from the Stripe API.

##### Pagination

Most APIs, when facing a large call, tend to return the results in pages. The CDK accommodates paging

> via the `next_page_token` function. This function is meant to extract the next page token from the latest

Reviewer: ```suggestion
via the `next_page_token` function. This function is meant to extract the next page "token" from the latest
```

> response. The CDK will continue making requests as long as the `next_page_function` continues returning

Reviewer: ```suggestion
response. The CDK will continue making requests as long as the `next_page_token` continues returning
```

> response. The CDK will continue making requests as long as the `next_page_function` continues returning

Reviewer: ```suggestion
response. The contents of a "token" are completely up to the developer: it can be an ID, a page number, a partial URL etc.. The CDK will continue making requests as long as the `next_page_token` function continues returning
```

> items. This can then be used in the `request_params` method to page through API responses. Here is an

Reviewer: ```suggestion
items. This can then be used in the `request_params` and other methods in `HttpStream` to page through API responses. Here is an
```

[example](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-stripe/source_stripe/source.py#L41) from the Stripe API.

##### Rate Limiting

> The CDK, by default, will conduct exponential backoff on the HTTP code 429 and unexpected exceptions,

Reviewer: ```suggestion
The CDK, by default, will conduct exponential backoff on the HTTP code 429 and any 5XX exceptions,
```

Note that Airbyte will always attempt to make as many requests as possible and only slow down if there are
errors. It is not currently possible to specify a rate limit Airbyte should adhere to when making requests.

### Practical Tips
* Airbyte recommends using the CDK template generator to develop with the CDK. The template generates
  created all the required scaffolding, with convenient TODOs, allowing developers to truly focus on
  implementing the API.

### Examples

Those interested in getting their hands dirty can check out implemeneted APIs:
* [Exchange Rates API](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-exchange-rates/source_exchange_rates/source.py) (Incremental)
* [Stripe API](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-stripe/source_stripe/source.py) (Incremental and Full-Refresh)
* [Slack API](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-slack/source_slack/source.py) (Incremental and Full-Refresh)

### Coming Soon
* Full OAuth 2.0 support
* Airbyte Java HTTP CDK

> * CDK for other protocols

Reviewer: ```suggestion
* CDK for Async HTTP endpoints (request-poll-wait style endpoints)
* CDK for other protocols
```

* updating the `request_params`, `path`, and other functions to incorporate slices

#### Secondary Features

The CDK offers other features that make writing HTTP APIs a breeze.

##### Authentication

The CDK supports token and OAuth2.0 authentication via the `TokenAuthenticator` and `Oauth2Authenticator` classes
respectively. Both authentication strategies are identical in that they place the api token in the `Authorization`
header. The `OAuth2Authenticator` goes an additional step further and has mechanisms to, given a refresh token,
refresh the current access token. Note that the `OAuth2Authenticator` currently only supports refresh tokens
and not the full OAuth2.0 loop.

Using either authenticator is as simple as passing the created authenticator into the relevant `HTTPStream`
constructor. Here is an [example](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-stripe/source_stripe/source.py#L242) from the Stripe API.

##### Pagination

Most APIs, when facing a large call, tend to return the results in pages. The CDK accommodates paging
via the `next_page_token` function. This function is meant to extract the next page "token" from the latest

> response. The CDK will continue making requests as long as the `next_page_token` continues returning

Reviewer: ```suggestion
response. The contents of a "token" are completely up to the developer: it can be an ID, a page number, a partial URL etc.. The CDK will continue making requests as long as the `next_page_token` function. The CDK will continue making requests as long as the `next_page_token` continues returning
```

> items. This can then be used in the `request_params` and other methods in `HttpStream` to page through API responses. Here is an

Reviewer: ```suggestion
non-`None` results. This can then be used in the `request_params` and other methods in `HttpStream` to page through API responses. Here is an
```
